File: frontend/app/api/auth/route.ts
// app/api/auth/route.ts — POST /api/auth
import { NextRequest, NextResponse } from 'next/server'
import { getDb } from '../../_shared/db'
import { createToken } from '../../_shared/auth'
import { SignJWT } from 'jose'
import crypto from 'crypto'

function sha256(text: string): string {
  return crypto.createHash('sha256').update(text).digest('hex')
}

function makeCode() {
  return Math.floor(100000 + Math.random() * 900000).toString()
}

export async function POST(request: NextRequest) {
  try {
    const body = await request.json()
    const { action, phone, code, password } = body as {
      action: string
      phone?: string
      code?: string
      password?: string
    }

    const db = getDb()

    if (action === 'send_code') {
      if (!phone || !/^1[3-9]\d{9}$/.test(phone)) {
        return NextResponse.json({ error: '手机号格式错误' }, { status: 400 })
      }
      const verificationCode = makeCode()
      const now = new Date()
      const expires = new Date(now.getTime() + 5 * 60 * 1000)
      db.prepare('DELETE FROM verification_codes WHERE phone=?').run(phone)
      db.prepare(
        'INSERT INTO verification_codes (phone, code, created_at, expires_at) VALUES (?, ?, ?, ?)'
      ).run(phone, verificationCode, now.toISOString(), expires.toISOString())
      console.log(`[DEV] Verification code for ${phone}: ${verificationCode}`)
      return NextResponse.json({ success: true, dev_code: verificationCode })
    }

    if (action === 'login') {
      if (!phone || !code) {
        return NextResponse.json({ error: '手机号和验证码必填' }, { status: 400 })
      }
      const row = db.prepare(
        'SELECT * FROM verification_codes WHERE phone=? AND used=0 ORDER BY created_at DESC LIMIT 1'
      ).get(phone) as any
      if (!row) return NextResponse.json({ error: '未找到有效验证码' }, { status: 400 })
      if (new Date() > new Date(row.expires_at)) {
        return NextResponse.json({ error: '验证码已过期，请重新获取' }, { status: 400 })
      }
      if (row.code !== code) {
        return NextResponse.json({ error: '验证码错误' }, { status: 400 })
      }
      db.prepare('UPDATE verification_codes SET used=1 WHERE id=?').run(row.id)
      let user = db.prepare('SELECT id FROM users WHERE phone=?').get(phone) as any
      let userId: number
      if (!user) {
        const result = db.prepare('INSERT INTO users (phone, created_at) VALUES (?, ?)').run(phone, new Date().toISOString())
        userId = result.lastInsertRowid as number
      } else {
        userId = user.id
      }
      const token = await createToken(userId, phone)
      return NextResponse.json({ token, user_id: userId })
    }

    if (action === 'login_password') {
      if (!phone || !password) {
        return NextResponse.json({ error: '手机号和密码必填' }, { status: 400 })
      }
      const pwHash = sha256(password)
      const user = db.prepare('SELECT id FROM users WHERE phone=? AND password_hash=?').get(phone, pwHash) as any
      if (!user) return NextResponse.json({ error: '手机号或密码错误' }, { status: 401 })
      const token = await createToken(user.id, phone)
      return NextResponse.json({ token, user_id: user.id })
    }

    if (action === 'register_password') {
      if (!phone || !/^1[3-9]\d{9}$/.test(phone)) {
        return NextResponse.json({ error: '手机号格式错误' }, { status: 400 })
      }
      if (!password || password.length < 6) {
        return NextResponse.json({ error: '密码至少6位' }, { status: 400 })
      }
      const existing = db.prepare('SELECT id FROM users WHERE phone=?').get(phone)
      if (existing) {
        return NextResponse.json({ error: '该手机号已注册，请直接登录' }, { status: 409 })
      }
      const pwHash = sha256(password)
      const result = db.prepare(
        'INSERT INTO users (phone, password_hash, created_at) VALUES (?, ?, ?)'
      ).run(phone, pwHash, new Date().toISOString())
      const userId = result.lastInsertRowid as number
      const token = await createToken(userId, phone)
      return NextResponse.json({ token, user_id: userId }, { status: 201 })
    }

    return NextResponse.json({ error: 'unknown action' }, { status: 400 })
  } catch (e: any) {
    return NextResponse.json({ error: e.message || 'server error' }, { status: 500 })
  }
}

File: api/auth/route.py
# api/auth/route.py — All auth endpoints: POST /api/auth
from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse
from .._shared.db import get_cursor
from .._shared.auth import create_token
import json, random, re
from datetime import datetime, timedelta

router = APIRouter()

def _json(body: bytes):
    return json.loads(body.decode())

@router.post("/auth")
def auth_endpoint(request: Request):
    body = json.loads(request.body.decode())
    action = body.get("action", "")

    if action == "send_code":
        return _send_code(body)
    elif action == "login":
        return _login(body)
    elif action == "login_password":
        return _login_password(body)
    elif action == "register_password":
        return _register_password(body)
    return JSONResponse({"error": "unknown action"}, status_code=400)

def _send_code(body: dict):
    phone = body.get("phone", "")
    if not re.match(r"^1[3-9]\d{9}$", phone):
        return JSONResponse({"error": "手机号格式错误"}, status_code=400)
    code = "".join(random.choices("0123456789", k=6))
    now = datetime.utcnow()
    expires = now + timedelta(minutes=5)
    with get_cursor() as c:
        c.execute("DELETE FROM verification_codes WHERE phone=?", (phone,))
        c.execute("INSERT INTO verification_codes (phone, code, created_at, expires_at) VALUES (?, ?, ?, ?)",
                   (phone, code, now.isoformat(), expires.isoformat()))
    print(f"[DEV] Verification code for {phone}: {code}")
    return {"success": True, "dev_code": code}

def _login(body: dict):
    phone = body.get("phone", "")
    code = body.get("code", "")
    if not phone or not code:
        return JSONResponse({"error": "手机号和验证码必填"}, status_code=400)
    with get_cursor() as c:
        c.execute("SELECT * FROM verification_codes WHERE phone=? AND used=0 ORDER BY created_at DESC LIMIT 1", (phone,))
        row = c.fetchone()
    if not row:
        return JSONResponse({"error": "未找到有效验证码"}, status_code=400)
    if datetime.utcnow() > datetime.fromisoformat(row["expires_at"]):
        return JSONResponse({"error": "验证码已过期，请重新获取"}, status_code=400)
    if row["code"] != code:
        return JSONResponse({"error": "验证码错误"}, status_code=400)
    with get_cursor() as c:
        c.execute("UPDATE verification_codes SET used=1 WHERE id=?", (row["id"],))
        c.execute("SELECT id FROM users WHERE phone=?", (phone,))
        user = c.fetchone()
        if not user:
            c.execute("INSERT INTO users (phone, created_at) VALUES (?, ?)", (phone, datetime.utcnow().isoformat()))
            user_id = c.lastrowid
        else:
            user_id = user["id"]
    token = create_token(user_id, phone)
    return {"token": token, "user_id": user_id}

def _login_password(body: dict):
    import hashlib
    phone = body.get("phone", "")
    password = body.get("password", "")
    if not phone or not password:
        return JSONResponse({"error": "手机号和密码必填"}, status_code=400)
    pw_hash = hashlib.sha256(password.encode()).hexdigest()
    with get_cursor() as c:
        c.execute("SELECT id FROM users WHERE phone=? AND password_hash=?", (phone, pw_hash))
        user = c.fetchone()
        if not user:
            return JSONResponse({"error": "手机号或密码错误"}, status_code=401)
        user_id = user["id"]
    token = create_token(user_id, phone)
    return {"token": token, "user_id": user_id}

def _register_password(body: dict):
    import hashlib
    phone = body.get("phone", "")
    password = body.get("password", "")
    if not re.match(r"^1[3-9]\d{9}$", phone):
        return JSONResponse({"error": "手机号格式错误"}, status_code=400)
    if len(password) < 6:
        return JSONResponse({"error": "密码至少6位"}, status_code=400)
    pw_hash = hashlib.sha256(password.encode()).hexdigest()
    with get_cursor() as c:
        c.execute("SELECT id FROM users WHERE phone=?", (phone,))
        if c.fetchone():
            return JSONResponse({"error": "该手机号已注册，请直接登录"}, status_code=409)
        c.execute("INSERT INTO users (phone, password_hash, created_at) VALUES (?, ?, ?)",
                   (phone, pw_hash, datetime.utcnow().isoformat()))
        user_id = c.lastrowid
    token = create_token(user_id, phone)
    return {"token": token, "user_id": user_id}
